source/common/grpc/async_client_manager_impl.cc

```diff
@@ -148,18 +149,58 @@ RawAsyncClientSharedPtr AsyncClientManagerImpl::getOrCreateRawAsyncClient(
   return client;
 }
 
-RawAsyncClientSharedPtr AsyncClientManagerImpl::RawAsyncClientCache::getCache(
-    const envoy::config::core::v3::GrpcService& config) const {
-  auto it = cache_.find(config);
-  if (it == cache_.end()) {
-    return nullptr;
-  }
-  return it->second;
+AsyncClientManagerImpl::RawAsyncClientCache::RawAsyncClientCache(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {
+  cache_eviction_timer_ = dispatcher.createTimer([this] { evictEntriesAndResetEvictionTimer(); });
 }
 
 void AsyncClientManagerImpl::RawAsyncClientCache::setCache(
     const envoy::config::core::v3::GrpcService& config, const RawAsyncClientSharedPtr& client) {
-  cache_[config] = client;
+  ASSERT(lru_map_.find(config) == lru_map_.end());
+  // Create a new cache entry at the beginning of the list.
+  lru_list_.emplace_front(config, client, dispatcher_.timeSource().monotonicTime());
+  lru_map_[config] = lru_list_.begin();
+  // If inserting to an empty cache, enable eviction timer.
+  if (lru_list_.size() == 1) {
+    evictEntriesAndResetEvictionTimer();
+  }
+}
+
+RawAsyncClientSharedPtr AsyncClientManagerImpl::RawAsyncClientCache::getCache(
+    const envoy::config::core::v3::GrpcService& config) {
+  RawAsyncClientSharedPtr client;
+  auto it = lru_map_.find(config);
+  if (it != lru_map_.end()) {
+    auto cache_entry = it->second;
+    // Reset the eviction timer if the next entry to expire is accessed.
+    bool should_reset_timer = (cache_entry == --lru_list_.end());
+    cache_entry->accessed_time_ = dispatcher_.timeSource().monotonicTime();
+    // Move the cache entry to the beginning of the list upon access.
+    lru_list_.splice(lru_list_.begin(), lru_list_, cache_entry);
+    // Get the cached async client before any cache eviction.
```

Comment:
Is the sequencing here important? I would think no since this entry is now the most recently used?

Reply:
It is theoretically possible that the most recently used entry get evicted, imagine a very slow machine that take 50 s to execute a command and already expire the most recently entry when it start the eviction method(though very unlikely)